sdk/Pulumi/Deployment/Deployment.cs

```diff
@@ -271,6 +280,68 @@ internal Task<bool> MonitorSupportsTransforms()
             return MonitorSupportsFeature("transforms");
         }
 
+        /// <summary>
+        /// Returns whether the resource monitor we are connected to supports the "invokeTransforms" feature across the RPC interface.
+        /// </summary>
+        public Task<bool> MonitorSupportsInvokeTransforms()
+        {
+            return MonitorSupportsFeature("invokeTransforms");
+        }
+
+        public async Task RegisterInvokeTransform(InvokeTransform transform)
+        {
+            var monitorSupportsInvokeTransforms = await MonitorSupportsInvokeTransforms().ConfigureAwait(false);
+            if (!monitorSupportsInvokeTransforms)
+            {
+                throw new InvalidOperationException("The Pulumi CLI does not support invoke transforms. Please update the Pulumi CLI.");
+            }
+
+            lock (_registrationLock)
+            {
+                _pendingRegistrations++;
+            }
+
+            var callbacks = await GetCallbacksAsync(CancellationToken.None).ConfigureAwait(false);
+            var callback = await AllocateInvokeTransform(callbacks.Callbacks, transform).ConfigureAwait(false);
+
+            await Monitor.RegisterStackInvokeTransform(callback).ConfigureAwait(false);
+            int updated;
+
+            var current = _registrationsComplete;
+            var shouldFlushBuffer = false;
+
+            lock (_registrationLock)
+            {
+                updated = --_pendingRegistrations;
+
+                if (updated == 0)
+                {
+                    shouldFlushBuffer = true;
+                    _registrationsComplete = new(TaskCreationOptions.RunContinuationsAsynchronously);
+                }
+            }
+
+            if (shouldFlushBuffer)
+            {
+                current.TrySetResult(true);
+            }
+        }
+
+        public async Task AwaitPendingRegistrations()
+        {
+            var awaitingPendingRegistrations = false;
+
+            lock (_registrationLock)
+            {
+                awaitingPendingRegistrations = _pendingRegistrations > 0;
```

Comment:
I think you can grab the _registrationsComplete.Task object within the lock as well.